### PR TITLE
[Merged by Bors] - TY-2852 Handle snippet too short for kpe

### PR DESCRIPTION
**Pull request by acrrd**
In some case the snippet is not long enough to perform kpe extraction, in this case we retry using the `title` concatenated with the `snippet`, if also this fails we use as the key phrase the `title` if not empty, otherwise the `snippet`.

### CODE REVIEW

File: discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs [hunk 1]
```diff
@@ -65,7 +65,7 @@ impl ValidMask {
 
     /// Checks if the valid mask is valid, i.e. at least `key_phrase_size` valid entries.
     pub(crate) fn is_valid(&self, key_phrase_size: usize) -> bool {
-        self.count() >= key_phrase_size
+        dbg!(dbg!(self.count()) >= dbg!(key_phrase_size))
```
Comment (rustonaut) [inline discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs:68]: :wrench: 

```suggestion
        self.count() >= key_phrase_size
```

File: discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs [hunk 2]
```diff
@@ -96,10 +96,11 @@ pub(crate) struct Encoding {
...
     pub(crate) fn is_valid(&self, vocab_size: usize, key_phrase_size: usize) -> bool {
-        self.token_ids.is_valid(vocab_size)
-            && self.attention_mask.is_valid()
-            && self.valid_mask.is_valid(key_phrase_size)
-            && self.active_mask.is_valid()
+        dbg!(self.active_mask.is_valid());
+        dbg!(self.token_ids.is_valid(vocab_size))
+            && dbg!(self.attention_mask.is_valid())
+            && dbg!(self.valid_mask.is_valid(key_phrase_size))
+            && dbg!(self.active_mask.is_valid())
```
Comment (rustonaut) [inline discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs:103]: :wrench:  more leftover `dbg!`s

File: discovery_engine_core/ai/ai/src/ranker/system.rs [hunk 3]
```diff
@@ -113,14 +113,26 @@ impl Ranker {
...
+                    .kpe
+                    .run(snippet)
+                    .or_else(|_| self.kpe.run(format!("{title} {snippet}")))
+                    .map(Into::into)
+                    .unwrap_or_else(|_| {
+                        vec![if !title.is_empty() {
+                            title.to_string()
+                        } else {
+                            snippet.to_string()
+                        }]
```
Comment (rustonaut) [inline discovery_engine_core/ai/ai/src/ranker/system.rs:134]: :question: Maybe we need some cleaning of the strings in this case?

**Review by rustonaut — CHANGES_REQUESTED** [review]

File: discovery_engine_core/ai/ai/src/ranker/system.rs (hunk 3 above)
Comment (acrrd) [reply discovery_engine_core/ai/ai/src/ranker/system.rs:134]: you mean something like `trim`? `clean_query` is called in another place using this as an input.

File: discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs (hunk 1 above)
Comment (acrrd) [reply discovery_engine_core/ai/kpe/src/tokenizer/encoding.rs:68]: Ah shit! :see_no_evil:

File: discovery_engine_core/ai/ai/src/ranker/system.rs (hunk 3 above)
Comment (rustonaut) [reply discovery_engine_core/ai/ai/src/ranker/system.rs:134]: I was actually thinking about cleaning not trimming as I wasn't sure if having "arbitrary input" as keywords could be a problem at some point in time. But then like you mentioned currently we clean it before using it.

**Review by rustonaut — APPROVED** [review]
lgtm

**Comment by acrrd** [comment]
bors merge

**Pull request closed without merging**